src/main/java/com/williamfiset/algorithms/datastructures/queue/IntQueue.java

```diff
@@ -5,50 +5,59 @@
  * is you need to know an upper bound on the number of elements that will be inside the queue at any
  * given time for this queue to work.
  *
- * @author William Fiset, william.alexandre.fiset@gmail.com
+ * @author William Fiset, william.alexandre.fiset@gmail.com, liujingkun, liujkon@gmail.com
  */
 package com.williamfiset.algorithms.datastructures.queue;
 
 public class IntQueue {
 
-  private int[] ar;
-  private int front, end, sz;
+  private int[] data;
+  private int front, end;
+  private int size;
 
   // maxSize is the maximum number of items
   // that can be in the queue at any given time
   public IntQueue(int maxSize) {
-    front = end = 0;
-    sz = maxSize + 1;
-    ar = new int[sz];
+    front = end = size = 0;
+    data = new int[maxSize];
   }
 
   // Return true/false on whether the queue is empty
   public boolean isEmpty() {
-    return front == end;
+    return size == 0;
   }
 
   // Return the number of elements inside the queue
   public int size() {
-    if (front > end) return (end + sz - front);
-    return end - front;
+    return size;
   }
 
   public int peek() {
-    return ar[front];
+    if (size == 0) {
+      throw new RuntimeException("Queue is empty");
+    }
+    front = front % data.length;
+    return data[front];
   }
 
   // Add an element to the queue
   public void enqueue(int value) {
-    ar[end] = value;
-    if (++end == sz) end = 0;
-    if (end == front) throw new RuntimeException("Queue too small!");
+    if (size == data.length) {
+      throw new RuntimeException("Queue too small!");
+    }
+    data[end++] = value;
+    size++;
+    end = end % data.length;
```

Comment:
ditto on the '%', maybe run tests to see if keeping to mod has a significant performance impact vs using '- data.length' or whatever equivalent.

Reply:
There are two solutions. One is to use the java conditional operator. Another is to validate the input of capacity to ensure it is 2^n and use bit manipulation to do the mod operation. which one we should use?

Reply:
Do whichever one you want, perhaps the 2^n approach would work well?

Reply:
I take the conditional operator finally for user experience, is it ok?